documentation/api/addAssertion.md

```diff
@@ -10,82 +10,237 @@ expect.addAssertion(pattern, handler);
+expect([1, 2, 3], 'to have item', 2);
+```
+
+The first parameter to `addAssertion` is a string or an array of strings
+describing the pattern(s) the assertion should match. The pattern takes the
+following structure:
+
+`<subject type> an assertion string <value type>`
+
+The words in angle brackets define what types the assertion applies to. These
+can be any of the internally-defined types or new types added via
+[expect.addType](../addType). In this example, the subject to `to have item`
+must be an array, while the value may be of any type.
+
+If mismatching types are used when an assertion is invoked, Unexpected throws an
+error with a helpful suggestion:
+
+```js
+expect('abcd', 'to have item', 'a');
+```
+
+```output
+expected 'abcd' to have item 'a'
+  The assertion does not have a matching signature for:
+    <string> to have item <string>
+  did you mean:
+    <array> to have item <any>
+```
+
+An assertion can only have a single assertion string, which must be
+provided after the subject type. This means that it's not possible to add
+assertions such as `<number> to be between <number> and <number>`.
+
+However, an assertion may define more than one value as long as there are
+no assertion strings in between the value-type definitions. The `to be between`
+assertion could instead be defined as:
+
+```js
+expect.addAssertion('<number> to be between <number> <number>', function(
+  expect,
+  subject,
+  value1,
+  value2
+) {
+  expect(subject, 'to be greater than', value1).and('to be less than', value2);
+});
+```
+
+```js
+expect(2, 'to be between', 1, 3);
+```
+
+## Alternations
+
+Different versions of the same assertion, or different assertions that share the
+same handler function, can be added using an array:
+
+<!-- unexpected-markdown freshExpect:true -->
 
 ```js
-var errorMode = 'default'; // use to control the error mode later in the example
 expect.addAssertion(
-  '<array> [not] to be (sorted|ordered) <function?>',
-  function(expect, subject, cmp) {
-    expect.errorMode = errorMode;
-    expect(subject, '[not] to equal', [].concat(subject).sort(cmp));
-  }
+  ['<array> to have item <any>', '<array> to have value <any>'],
+  handler
 );
 ```
 
-The above assertion definition makes the following expects possible:
+```js
+expect([1, 2, 3], 'to have item', 2);
+expect([1, 2, 3], 'to have value', 3);
+```
+
+However, array patterns are ideal for assertions that have different types for
+the subject or values; for instance, if the second assertion was defined as
+`<array> to have value <number>`. When the difference is only in the assertion
+string, an alternation is more handy:
+
+<!-- unexpected-markdown freshExpect:true -->
 
 ```js
-expect([1, 2, 3], 'to be sorted');
-expect([1, 2, 3], 'to be ordered');
-expect([2, 1, 3], 'not to be sorted');
-expect([2, 1, 3], 'not to be ordered');
-expect([3, 2, 1], 'to be sorted', function(x, y) {
-  return y - x;
+expect.addAssertion('<array> to have (item|value) <any>', handler);
```

Comment:
@alexjeffburke, this is what I meant when I mentioned that the handler function is referenced in examples that follow. This `handler` on this line is the function declared on line 21. 

If you'd like, I can rewrite the first example assertion on line 26 to this:

```js
expect.addAssertion('<array> to have item <any>', function handler() {
  expect(subject, 'to contain', value);
});
```

That way, the handler is inlined and named, and can therefore be referenced later on. I'm just afraid that might be hard to spot and therefore, a bit confusing. 

Alternatively, I could copy/paste the handler function here and other places where it's reused, I don't think it's too big of a deal.

Let me know what you prefer.

Reply:
Go with your inlining idea :) I don’t wanna block a PR that improves things for something that is ultimately minor.

Reply:
It didn't work, I went with copy/pasting instead.